airo-robots/airo_robots/drives/hardware/kelo_robile.py

```diff
@@ -0,0 +1,119 @@
+from airo_tulip.api.client import KELORobile as KELORobileClient  # type: ignore
+from airo_tulip.hardware.platform_driver import PlatformDriverType  # type: ignore
+from airo_typing import Vector3DType
+
+
+class KELORobile(MobileRobot):
+    """KELO Robile platform implementation of the MobileRobot interface.
+
+    The KELO Robile platform consists of several drives with castor wheels which are connected via EtherCAT and can
+    be controlled individually. The airo-tulip API, which is used here, provides a higher level interface which
+    controls the entire platform."""
+
+    def __init__(self, robot_ip: str, robot_port: int = 49789):
+        """Connect to the KELO robot.
+
+        The KELO robot should already be running the airo-tulip server. If this is not the case, any messages
+        sent from the client may be queued and executed once the server is started, resulting in unexpected
+        and/or sudden movements.
+
+        Args:
+            robot_ip: IP address of the KELO CPU brick.
+            robot_port: Port to connect on (default: 49789)."""
+        self._kelo_robile = KELORobileClient(robot_ip, robot_port)
+
+    def align_drives(self, x: float, y: float, a: float, timeout: float = 1.0) -> AwaitableAction:
+        """Align all drives for driving in a direction given by the linear and angular velocities.
+
+        Beware that sending any other velocity commands before the awaitable is done may cause unexpected behaviour,
+        as this affects the "aligned" condition.
+
+        Args:
+            x: The x velocity (linear, m/s).
+            y: The y velocity (linear, m/s)
+            a: The angular velocity (rad/s).
+            timeout: The awaitable will finish after this time at the latest.
+
+        Returns:
+            An AwaitableAction which will check if the drives are aligned, or if the timeout has expired."""
+        self._kelo_robile.align_drives(x, y, a, timeout=timeout)
+
+        action_sent_time = time.time_ns()
+        return AwaitableAction(
+            lambda: self._kelo_robile.are_drives_aligned() or time.time_ns() - action_sent_time > timeout * 1e9,
+            default_timeout=2 * timeout,
+            default_sleep_resolution=0.002,
+        )
+
+    def set_platform_velocity_target(self, x: float, y: float, a: float, timeout: float) -> AwaitableAction:
+        self._kelo_robile.set_platform_velocity_target(x, y, a, timeout=timeout)
+
+        action_sent_time = time.time_ns()
+        return AwaitableAction(
+            lambda: time.time_ns() - action_sent_time > timeout * 1e9,
+            default_timeout=2 * timeout,
+            default_sleep_resolution=0.002,
+        )
+
+    def move_platform_to_pose(self, x: float, y: float, a: float, timeout: float) -> AwaitableAction:
+        target_pose = np.array([x, y, a])
+        action_start_time = time.time_ns()
+        action_timeout_time = action_start_time + timeout * 1e9
+
+        def control_loop() -> bool:
```

Comment:
Because `AwaitableAction` busy-waits, using `move_platform_to_pose` in parallel with other `AwaitableAction`s, e.g., to move the UR, does not work as expected.

For example,

```python
q = arm.get_joint_configuration().copy()
q[0] += np.pi / 8

gripper_awaitable = gripper.open()
arm_awaitable = arm.move_to_joint_configuration(q)
platform_awaitable = platform.move_platform_to_pose(-0.2, 0.0, 0.0, 10.0)

platform_awaitable.wait()
gripper_awaitable.wait()
arm_awaitable.wait()
```

will move the platform, gripper, and arm at the same time, as one would expect. However,

```python
q = arm.get_joint_configuration().copy()
q[0] += np.pi / 8

gripper_awaitable = gripper.open()
arm_awaitable = arm.move_to_joint_configuration(q)
platform_awaitable = platform.move_platform_to_pose(-0.2, 0.0, 0.0, 10.0)

# This order has changed
gripper_awaitable.wait()
arm_awaitable.wait()
platform_awaitable.wait()
```

will only move the platform once the gripper and arm `AwaitableAction`s have finished.

This is because the control loop is implemented as a function to be passed to the `AwaitableAction`, which only starts executing once `wait()` is called.

The solution is to move the control loop to `airo-tulip` and send a command, similarly to how the other methods of `KELORobile` work. Then, the KELO itself can handle the execution of the loop, and the program flow of the application code using `airo-robots` can continue, achieving the asynchronous behaviour we expect.

Reply:
executing the control loop in the awaitable action is indeed a bit unfortunate I think, the wait loop should implement a (lightweight) status checker. Any long-term control loop should be executed a separate thread in the caller process or (preferably?) be handled by the (remote) robot server process

Reply:
> Any long-term control loop should be executed a separate thread in the caller process or (preferably?) be handled by the (remote) robot server process

I agree that this should preferably be handled by the robot server process. I ran into some issues however and decided to park it for now and implement this the easy way. See below

A first attempt to handle this in the robot server process without needing to write a new `PlatformController` was made, with a separate thread running the position control loop from within the `TulipServer` (both classes in `airo-tulip`). However, accessing the platform driver from a separate thread made the KELO drives fail silently.

For now, I have resorted to running the thread from within `airo-robots` (See commit [a0e94f5](https://github.com/airo-ugent/airo-mono/pull/147/commits/a0e94f5003d2f23edf2288ab088267e97a351f98)). We could raise an issue for this to be moved to `airo-tulip`, but for now I consider this solved.